### PR TITLE
Allow omega to take prefixing PyTree to handle buffer edge case

### PR DESCRIPTION
**Pull request by RaderJason**
`jtu.tree_map(fn, tree, *rest)` uses `tree` to determine where to evaluate
`f(tree, *rest)`. `tree` must be a prefixing PyTree to
every PyTree in  `*rest`.

This adds support for prefixing trees. in particular:
```
ω(a, b).call(fn) = jax.tree_util(lambda x, y: fn(y), b, a)
```
and
```
ω(a).call(fn) = ω(a, a).call(fn)
```

This was included to handle buffers in the equinox internal while loop,
where we may wish to use ω to index into buffers, rather than into the PyTrees
resulting from wrapping subtrees in buffers.

### CODE REVIEW

File: equinox/internal/_loop/common.py [hunk 1]
```diff
@@ -61,6 +61,17 @@ class _BufferItem(Module):
...
+        return self._buffer._set(pred, self._item, self._buffer[self._item] + x)
+
+    def multiply(self, x, *, pred=True):
+        return self._buffer._set(pred, self._item, self._buffer[self._item] * x)
+
+    def divide(self, x, *, pred=True):
+        return self._buffer._set(pred, self._item, self._buffer[self._item] / x)
+
+    def power(self, x, *, pred=True):
+        return self._buffer._set(pred, self._item, self._buffer[self._item] ** x)
```
Comment (patrick-kidger) [inline equinox/internal/_loop/common.py:74]: Let's make it so that e.g. `add` lowers to `array.at[...].add`.

Probably replace `_set(self, pred, item, x) -> _op(self, pred, item, x , op)`, where `op` is a string indicating which method to call: `getattr(array.at[...], op)`.

File: equinox/internal/_omega.py [hunk 2]
```diff
@@ -29,13 +29,23 @@ class _ω(metaclass=_Metaω):
...
         This is entirely equivalent to the above.
+
+    !!! edge case
+
+        If if a tuple `(a, b)` is passed then
+        ```
+        ω(a,b).call(fn) = jax.tree_util.tree_map(lambda x,y: fn(y), b, a)
+        ```
+        where b is a prefix PyTree of a. This evaluates a at the nodes corresponding
+        to the leaves of b.
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:40]: This isn't an edge case, this is the intended behaviour!

Also, the label after the `!!!` has specific meaning: https://squidfunk.github.io/mkdocs-material/reference/admonitions/#supported-types

File: equinox/internal/_omega.py [hunk 3]
```diff
@@ -29,13 +29,23 @@ class _ω(metaclass=_Metaω):
...
     """
 
-    def __init__(self, value, is_leaf=None):
+    def __init__(self, *value, is_leaf=None):
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:43]: I'd suggest making this be slightly less general: `self, value, *, structure=sentinel, is_leaf=None`

File: equinox/internal/_omega.py [hunk 4]
```diff
@@ -227,6 +260,7 @@ def fn(self, other):
 for (name, op) in [
     ("set", lambda x, y, z, **kwargs: x.at[y].set(z, **kwargs)),
     ("add", lambda x, y, z, **kwargs: x.at[y].add(z, **kwargs)),
+    ("sub", lambda x, y, z, **kwargs: x.at[y].add(-z, **kwargs)),
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:263]: I wouldn't include this; we should match the `ndarray.at` API.

File: equinox/internal/_loop/common.py [hunk 5]
```diff
@@ -20,14 +20,24 @@ class _Buffer(Module):
...
-    def _set(self, pred, item, x):
+    # def _set(self, pred, item, x):
+    #     pred = pred & self._pred
+    #     if isinstance(self._array, _Buffer):
+    #         array = self._array._set(pred, item, x)
+    #     else:
+    #         old_x = self._array[item]
+    #         x = jnp.where(pred, x, old_x)
+    #         array = self._array.at[item].set(x)
+    #     return _Buffer(array, self._pred, self._tag)
```
Comment (patrick-kidger) [inline equinox/internal/_loop/common.py:31]: ?

File: equinox/internal/_omega.py [hunk 6]
```diff
@@ -29,13 +31,22 @@ class _ω(metaclass=_Metaω):
         ```
 
         This is entirely equivalent to the above.
+
+
+        If if a tuple `(a, b)` is passed then
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:36]: Not a tuple.

File: equinox/internal/_omega.py [hunk 7]
```diff
@@ -44,27 +55,39 @@ def __init__(self, value, is_leaf=None):
...
         self.is_leaf = is_leaf
 
     def __repr__(self):
         return f"ω({self.ω})"
 
     def __getitem__(self, item):
         return ω(
-            jtu.tree_map(lambda x: x[item], self.ω, is_leaf=self.is_leaf),
+            jtu.tree_map(
+                lambda x, y: y[item], self.struct, self.ω, is_leaf=self.is_leaf
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:73]: underscore for unused argument

File: equinox/internal/_omega.py [hunk 8]
```diff
@@ -44,27 +55,39 @@ def __init__(self, value, is_leaf=None):
...
+            ),
+            structure=self.struct,
             is_leaf=self.is_leaf,
         )
 
     def call(self, fn):
         return ω(
-            jtu.tree_map(fn, self.ω, is_leaf=self.is_leaf),
+            jtu.tree_map(
+                lambda x, y: fn(y[...]), self.struct, self.ω, is_leaf=self.is_leaf
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:82]: Remove the `[...]`

File: equinox/internal/_loop/common.py (hunk 5 above)
Comment (RaderJason) [reply equinox/internal/_loop/common.py:31]: oops!

File: equinox/internal/_omega.py [hunk 9]
```diff
@@ -29,13 +31,22 @@ class _ω(metaclass=_Metaω):
         ```
 
         This is entirely equivalent to the above.
+
+
+        If `structure` is passed then
+        ```
+        ω(a,structure=b).call(fn) = jax.tree_util.tree_map(lambda x,y: fn(y), b, a)
+        ```
+        where b is a prefix PyTree of a. This evaluates a at the nodes corresponding
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:40]: Nits: add backticks around variable names. Add spaces to code block to format it like Black would. Rename `x` to `_`.

File: equinox/internal/_omega.py [hunk 10]
```diff
@@ -29,13 +31,22 @@ class _ω(metaclass=_Metaω):
...
+        to the leaves of b.
     """
 
-    def __init__(self, value, is_leaf=None):
+    def __init__(self, value, *, structure=sentinel, is_leaf=None):
         """
         **Arguments:**
 
-        - `value`: The PyTree to wrap.
+        - `value`: The PyTree to wrap. If value is a two tuple then the first
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:48]: Revert the change to this line

File: equinox/internal/_omega.py [hunk 11]
```diff
@@ -44,27 +55,39 @@ def __init__(self, value, is_leaf=None):
...
         self.is_leaf = is_leaf
 
     def __repr__(self):
         return f"ω({self.ω})"
 
     def __getitem__(self, item):
         return ω(
-            jtu.tree_map(lambda x: x[item], self.ω, is_leaf=self.is_leaf),
+            jtu.tree_map(
+                lambda _x, y: y[item], self.struct, self.ω, is_leaf=self.is_leaf
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:73]: Rename `_x` to just `_`.

File: equinox/internal/_omega.py [hunk 12]
```diff
@@ -91,17 +114,30 @@ def _equal_code(fn1: Optional[Callable], fn2: Optional[Callable]):
...
         if isinstance(other, ω):
-            if jtu.tree_structure(self.ω) != jtu.tree_structure(other.ω):
+            if jtu.tree_structure(self.struct) != jtu.tree_structure(other.struct):
                 raise ValueError("PyTree structures must match.")
             if not _equal_code(self.is_leaf, other.is_leaf):
                 raise ValueError("`is_leaf` must match.")
             return ω(
-                jtu.tree_map(op, self.ω, other.ω, is_leaf=self.is_leaf),
+                jtu.tree_map(
+                    lambda _x, y, z: op(y[...], z[...]),
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:123]: Superfluous `[...]` again.

File: equinox/internal/_omega.py [hunk 13]
```diff
@@ -241,7 +241,7 @@ def fn(self, other):
                 raise ValueError("is_leaf specifications must match.")
             return ω(
                 jtu.tree_map(
-                    lambda _x, y, z: op(y, self.item, z[...]),
+                    lambda _, y, z: op(y, self.item, z[...]),
```
Comment (patrick-kidger) [inline equinox/internal/_omega.py:244]: You still have several needless `[...]`.

**Comment by patrick-kidger** [comment]
Closing as discussed offline.
